paymentsheet/src/main/java/com/stripe/android/paymentsheet/paymentdatacollection/polling/PollingViewModel.kt

```diff
@@ -0,0 +1,188 @@
+package com.stripe.android.paymentsheet.paymentdatacollection.polling
+
+import android.os.SystemClock
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.stripe.android.core.injection.DUMMY_INJECTOR_KEY
+import com.stripe.android.core.injection.InjectorKey
+import com.stripe.android.model.StripeIntent
+import com.stripe.android.polling.IntentStatusPoller
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.ZERO
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+private const val KEY_CURRENT_POLLING_START_TIME = "KEY_CURRENT_POLLING_START_TIME"
+
+internal interface TimeProvider {
+    fun currentTimeInMillis(): Long
+}
+
+internal class DefaultTimeProvider @Inject constructor() : TimeProvider {
+    override fun currentTimeInMillis(): Long {
+        return SystemClock.elapsedRealtime()
```

Comment:
`SystemClock` is part of the Android framework. For unit tests, I added a `FakeTimeProvider` that instead uses a simple call to `System.currentTimeMillis()`. The latter is not suitable for production because its value can be changed by the user and the phone network.